exotations/solvers/exotica_ddp_solver/src/feasibility_driven_ddp_solver.cpp

```diff
@@ -0,0 +1,623 @@
+
+    AllocateData();
+    auto X_warm = prob_->get_X();
+    X_warm.col(0) = prob_->ApplyStartState();  // Apply start state
+    auto U_warm = prob_->get_U();
+    for (int t = 0; t < T_ - 1; ++t)
+    {
+        xs_[t] = X_warm.col(t);
+        us_[t] = U_warm.col(t);
+    }
+    xs_.back() = X_warm.col(T_ - 1);
+    is_feasible_ = false;  // We assume the first iteration is always infeasible. TODO: Make this configurable
+
+    prob_->ResetCostEvolution(GetNumberOfMaxIterations() + 1);
+    prob_->PreUpdate();
+    solution.resize(T_ - 1, NU_);
+
+    // Initial roll-out to get initial cost
+    cost_ = 0.0;
+    for (int t = 0; t < T_ - 1; ++t)
+    {
+        prob_->Update(xs_[t], us_[t], t);
+        cost_ += dt_ * (prob_->GetStateCost(t) + prob_->GetControlCost(t));
+    }
+    // Reset shooting nodes so we can warm-start from state trajectory
+    prob_->set_X(X_warm);
+    cost_ += prob_->GetStateCost(T_ - 1);
+    prob_->SetCostEvolution(0, cost_);
+
+    xreg_ = std::max(regmin_, parameters_.RegularizationRate);
+    ureg_ = std::max(regmin_, parameters_.RegularizationRate);
+    was_feasible_ = false;
+
+    bool diverged = false;
+    bool converged = false;
+
+    bool recalcDiff = true;
+    int iter;
+    for (iter = 1; iter <= GetNumberOfMaxIterations(); ++iter)
+    {
+        // Check whether user interrupted (Ctrl+C)
+        if (Server::IsRos() && !ros::ok())
+        {
+            if (debug_) HIGHLIGHT("Solving cancelled by user");
+            prob_->termination_criterion = TerminationCriterion::UserDefined;
+            break;
+        }
+
+        backward_pass_timer.Reset();
+        while (true)
+        {
+            try
+            {
+                ComputeDirection(recalcDiff);
+            }
+            catch (std::exception& e)
+            {
+                // HIGHLIGHT("Increase regularization in ComputeDirection and try again." << e.what())
+                recalcDiff = false;
+                IncreaseRegularization();
+                if (xreg_ == regmax_)
+                {
+                    WARNING_NAMED("FeasibilityDrivenDDPSolver::Solve", "State regularization exceeds maximum regularization: " << xreg_ << " > " << regmax_)
+                    diverged = true;
+                    break;
+                }
+                else
+                {
+                    continue;
+                }
+            }
+            break;
+        }
+        time_taken_backward_pass_ = backward_pass_timer.GetDuration();
+
+        if (diverged)
+        {
+            WARNING("Terminating: Divergence in ComputeDirection / BackwardPass.");
+            break;
+        }
+
+        UpdateExpectedImprovement();
+
+        // We need to recalculate the derivatives when the step length passes
+        recalcDiff = false;
+        line_search_timer.Reset();
+        for (int ai = 0; ai < alpha_space_.size(); ++ai)
+        {
+            steplength_ = alpha_space_(ai);
+
+            try
```

Comment:
If used in a loop for flow control, use a variable/if/else instead of try/catch. Unlike error handling, this actually impacts the performance depending on often the catch gets called.
Relevant lines: 169, 462, 466, 484, 600, 604, 615